SecondBridge/SecondBridge/Data/BinarySearchTree.swift

```diff
@@ -0,0 +1,236 @@
+    */
+    func get(part: PartsTree) -> BinarySearchTree {
+        
+        switch self {
+        case .Empty:
+            return .Empty
+        case let .Node(_, left: _, right: rigth) where part == .Right:
+            return rigth
+        case let .Node(_, left: left, right: _) where part == .Left:
+            return left
+        default:
+            return self
+        }
+    }
+    
+    /**
+    * Returns a new BinarySearchTree with the current contents and the provided item.
+    */
+    func add(element:T) -> BinarySearchTree {
+        switch self {
+        case .Empty:
+            return BinarySearchTree.Node(element, left: BinarySearchTree.Empty, right: BinarySearchTree.Empty)
+        case let .Node(value, left, right):
+            if element > value {
+                return BinarySearchTree.Node(value, left: left, right: right.add(element))
+            } else {
+                return BinarySearchTree.Node(value, left: left.add(element), right: right)
+            }
+        }
+    }
+    
+    /**
+    Checks if a certain element is included in the current BST.
+    
+    - parameter element: The element to be checked.
+    
+    - returns: True if the BST contains this element.
+
+    */
+    func search(element:T) -> Bool{
+        
+        switch self{
+        case .Empty:
+            return false
+        case let .Node(value, _, _) where element == value:
+            return true
+        case let .Node(value, left, _) where value > element:
+            return left.search(element)
+        case let .Node(value, _, right) where value < element:
+            return right.search(element)
+        default:
+            print("Error")
+        }
+        return false
+    }
+    
+    /**
+    *  Removes the provided element from the current BST, and returns a new BST without that element
+    */
+    func remove(element:T) -> BinarySearchTree? {
+        
+        switch self {
+        case .Empty:
+            return self
+        case let .Node(value, left, right) where value > element:
+            return BinarySearchTree.Node(value, left: left.remove(element)!, right: right)
+        case let .Node(value, left, right) where value < element:
+            return BinarySearchTree.Node(value, left:left, right: right.remove(element)!)
+        case let .Node(value, .Empty, .Empty) where value == element:
+            return BinarySearchTree.Empty
+        case let .Node(value, left, .Empty) where value == element:
+            return left
+        case let .Node(value, .Empty, right) where value == element:
+            return right
+        case let .Node(value, left, right) where value == element:
+            let a = findMin(right)
+            switch a {
+            case let .Node(value, _,_):
+                return BinarySearchTree.Node(value, left: left, right: right.remove(value)!)
+            default:
+                return .Empty
+            }
+        default:
+            print("Error")
```

Comment:
As in a previous function, this is an unnecessary side-effect from the user's perspective IMO.